exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java

```diff
@@ -361,6 +365,102 @@ public FileMetadata getFileMetadata(Path location) {
+    Multimap<Path, T> metadataMultimap = LinkedListMultimap.create();
+    metadata.forEach((key, value) -> metadataMultimap.put(key.getParent(), value));
+
+    Map<Path, SegmentMetadata> result = new HashMap<>();
+    metadataMultimap.asMap().forEach((key, value) -> result.put(key, combineToSegmentMetadata(value, column)));
+
+    return result;
+  }
+
+  private static Map<Path, SegmentMetadata> getMetadataForSegments(Map<Path, SegmentMetadata> metadata, SchemaPath column) {
+    Multimap<Path, SegmentMetadata> metadataMultimap = LinkedListMultimap.create();
+    metadata.forEach((key, value) -> metadataMultimap.put(key.getParent(), value));
+
+    Map<Path, SegmentMetadata> result = new HashMap<>();
+    metadataMultimap.asMap().forEach((key, value) -> result.put(key, combineSegmentMetadata(value, column)));
+
+    return result;
+  }
+
+  private static <T extends BaseMetadata & LocationProvider> SegmentMetadata combineToSegmentMetadata(Collection<T> metadataList, SchemaPath column) {
+    List<Path> metadataLocations = metadataList.stream()
+        .map(metadata -> metadata.getPath()) // used lambda instead of method reference due to JDK-8141508
+        .collect(Collectors.toList());
+    return combineToSegmentMetadata(metadataList, column, metadataLocations);
+  }
+
+  private static SegmentMetadata combineSegmentMetadata(Collection<SegmentMetadata> metadataList, SchemaPath column) {
+    List<Path> metadataLocations = metadataList.stream()
+        .flatMap(metadata -> metadata.getLocations().stream())
+        .collect(Collectors.toList());
+
+    return combineToSegmentMetadata(metadataList, column, metadataLocations);
+  }
+
+  /**
+   * Returns {@link SegmentMetadata} which is combined metadata of list of specified metadata
+   *
+   * @param metadataList      metadata to combine
+   * @param column            segment column
+   * @param metadataLocations locations of metadata combined in resulting segment
+   * @param <T>               type of metadata to combine
+   * @return {@link SegmentMetadata} from combined metadata
+   */
+  private static <T extends BaseMetadata & LocationProvider> SegmentMetadata combineToSegmentMetadata(Collection<T> metadataList,
+      SchemaPath column, List<Path> metadataLocations) {
+    List<StatisticsHolder> segmentStatistics =
+        Collections.singletonList(
+            new StatisticsHolder<>(
+                TableStatisticsKind.ROW_COUNT.mergeStatistics(metadataList),
+                TableStatisticsKind.ROW_COUNT));
+    // this code is used only to collect segment metadata to be used only during filtering,
+    // so metadata identifier is not required here and in other places in this class
+    MetadataInfo metadataInfo = new MetadataInfo(MetadataType.SEGMENT, MetadataInfo.GENERAL_INFO_KEY, null);
+    T firstMetadata = metadataList.iterator().next();
+
+    return new SegmentMetadata(firstMetadata.getTableInfo(), metadataInfo, column, firstMetadata.getSchema(),
```

Comment:
It's 10 arguments constructor. Maybe it's time for builder ? :)

Reply:
Agree, done.